x/ccv/provider/keeper/throttle.go

```diff
@@ -0,0 +1,420 @@
+package keeper
+
+import (
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	providertypes "github.com/cosmos/interchain-security/x/ccv/provider/types"
+	ccvtypes "github.com/cosmos/interchain-security/x/ccv/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+// This file contains functionality relevant to the throttling of slash and vsc matured packets, aka circuit breaker logic.
+
+// HandlePendingSlashPackets handles all or some portion of pending slash packets received by consumer chains,
+// depending on throttling logic. The slash meter is decremented appropriately in this method.
+func (k Keeper) HandlePendingSlashPackets(ctx sdktypes.Context) {
+
+	meter := k.GetSlashMeter(ctx)
+
+	// Don't start iterating if meter is 0 or negative in value
+	if !meter.IsPositive() {
+		return
+	}
+
+	handledEntries := []providertypes.SlashPacketEntry{}
+
+	// Iterate through ordered (by received time) slash packet entries from any consumer chain
+	k.IteratePendingSlashPacketEntries(ctx, func(entry providertypes.SlashPacketEntry) (stop bool) {
+
+		// Obtain the validator power relevant to the slash packet that's about to be handled
+		// (this power will be removed via jailing or tombstoning)
+		valPower := k.stakingKeeper.GetLastValidatorPower(ctx, entry.ValAddr)
```

Comment:
I have a weird deja vu that I already mentioned this elsewhere, so apologies if I have!

The entry.ValAddr is the consensus address but the validator power is stored using the operator address so this query is not going to work properly.

Reply:
I would suggest naming all the places where ValAddress or ConsAddress is used (expliclity, by type), or implicitly ([]byte) 'operatorAddr' or 'consAddr' or similar to avoid this kind of confusion.

Reply:
Great find, I believe this isn't caught by e2e or unit tests because we populate those bytes ourselves. Glad diff tests came in handy! I believe integration tests around throttling would have caught this bug as well